src/main/java/net/ripe/ipresource/ImmutableResourceSet.java

```diff
@@ -0,0 +1,363 @@
+        Entry<UniqueIpResource, IpResource> potentialMatch = resourcesByEndPoint.ceilingEntry(resource.getStart());
+        return potentialMatch != null && potentialMatch.getValue().contains(resource);
+    }
+
+    public boolean contains(ImmutableResourceSet other) {
+        for (IpResource resource: other) {
+            if (!contains(resource)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public boolean containsType(IpResourceType type) {
+        for (IpResource resource: resourcesByEndPoint.values()) {
+            if (type == resource.getType()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean intersects(IpResource resource) {
+        Entry<UniqueIpResource, IpResource> potentialMatch = resourcesByEndPoint.ceilingEntry(resource.getStart());
+        return potentialMatch != null && potentialMatch.getValue().overlaps(resource);
+    }
+
+    public boolean intersects(ImmutableResourceSet that) {
+        if (this.isEmpty() || that.isEmpty()) {
+            return false;
+        }
+        Iterator<IpResource> thisIterator = this.iterator();
+        Iterator<IpResource> thatIterator = that.iterator();
+        IpResource thisResource = thisIterator.next();
+        IpResource thatResource = thatIterator.next();
+        while (thisResource != null && thatResource != null) {
+            if (thisResource.overlaps(thatResource)) {
+                return true;
+            }
+            int compareTo = thisResource.getEnd().compareTo(thatResource.getEnd());
+            if (compareTo <= 0) {
+                thisResource = thisIterator.hasNext() ? thisIterator.next() : null;
+            }
+            if (compareTo >= 0) {
+                thatResource = thatIterator.hasNext() ? thatIterator.next() : null;
+            }
+        }
+        return false;
+    }
+
+    public static ImmutableResourceSet parse(String s) {
+        String[] resources = s.split(",");
+        Builder builder = new Builder();
+        for (String r : resources) {
+            String trimmed = r.trim();
+            if (!trimmed.isEmpty()) {
+                builder.add(IpResource.parse(trimmed));
+            }
+        }
+        return builder.build();
+    }
+
+    @Override
+    public String toString() {
+        return resourcesByEndPoint.values().stream().map(Objects::toString).collect(Collectors.joining(", "));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (! (obj instanceof ImmutableResourceSet)) {
+            return false;
+        }
+        ImmutableResourceSet other = (ImmutableResourceSet) obj;
+        return resourcesByEndPoint.equals(other.resourcesByEndPoint);
+    }
+
+    @Override
+    public int hashCode() {
+        return resourcesByEndPoint.hashCode();
+    }
+
+    public static class Builder {
```

Comment:
Do we need a separate Builder here or would it be possible to use `IpResourceSet`  for that?

Reply:
The ownership of the interlan `resourcesByEndPoint` tree is transferred by the `Builder` to the `ImmutableResourceSet`, so that's why I didn't use the `IpResourceSet` here. I've added checks to ensure the builder is only used once.

It is possible to use `IpResourceSet` internally from here, but then we'd have to "steal" the `resourcesByEndPoint` tree. This would removes some code duplication.